aesd-char-driver/main.c

```diff
@@ -0,0 +1,396 @@
+
+
+static void * kernel_mem_realloc(void* old_mem, size_t old_len,
+                   size_t new_len, unsigned int mode)
+{
+    void *new_mem;
+    new_mem = kmalloc(new_len, mode);
+    if (new_mem == NULL) {
+        printk(KERN_ERR "aesdchar: kernel_mem_realloc allocation error %d \n", __LINE__);
+        return NULL;
+    }
+    else {
+        memcpy(new_mem, old_mem, old_len);
+        kfree(old_mem);
+    }
+
+    return new_mem;
+}
+
+static int aesd_open(struct inode *inode, struct file *filp)
+{
+	struct aesd_dev *dev; /* device information */
+    PDEBUG("open \n");    
+	dev = container_of(inode->i_cdev, struct aesd_dev, cdev);
+	filp->private_data = dev; /* for other methods */
+    return 0;
+}
+
+static int aesd_release(struct inode *inode, struct file *filp)
+{
+	struct aesd_dev *dev; /* device information */
+    int ret = 0;
+    PDEBUG("release \n");
+    dev = (struct aesd_dev *)filp->private_data;
+	if (dev == NULL)
+		ret = -ENODEV;    
+    return ret;
+}
+
+static ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    unsigned long copy_size;    
+    ssize_t rc = 0;
+    struct aesd_buffer_entry *dev_buf;
+    size_t entry_offset_byte = 0;
+    int ret = 0;
+    struct aesd_dev *dev_struct = (struct aesd_dev *)filp->private_data;
+
+    PDEBUG("read %zu bytes with offset %lld \n",count,*f_pos);
+    /**
+     * TODO: handle read
+     */
+ 
+    ret = mutex_lock_interruptible(&dev_struct->aesdchar_mutex);
+    if (ret < 0) {
+        printk(KERN_ERR"Mutex lock failed \n");
+        return ret;
+    }
+
+    dev_buf = aesd_circular_buffer_find_entry_offset_for_fpos(dev_struct->aesd_buffer, 
+                    *f_pos, &entry_offset_byte);
+
+    if(dev_buf == NULL) {
+        rc = 0;
+        goto return_no_mem;
+    }
+    
+    copy_size = copy_to_user(buf, dev_buf->buffptr, dev_buf->size);
+    if ( copy_size != 0) {        
+        PDEBUG("Copy_to_user error \n");
+        rc = -EFAULT;
+        goto err_cleanup;
+    }
+
+    *f_pos += dev_buf->size; 
+    err_cleanup:
+    if (rc == 0) {
+        rc = dev_buf->size;
+    }
+    return_no_mem:
+	mutex_unlock(&dev_struct->aesdchar_mutex);
+
+    return rc;
+}
+
+
+static ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
+                loff_t *f_pos)
+{    
+    unsigned long copy_size;
+    ssize_t rc = 0;
+    int ret = 0;
+    char *buf_ptr = NULL;
+    struct aesd_buffer_entry *aesd_buf_entry;
+    char *nl_index = NULL;    
+    int status = 0;    
+    char * device_buffer = NULL;
+    struct aesd_dev *dev_struct =  (struct aesd_dev *)filp->private_data;
+
+    PDEBUG("write %zu bytes with offset %lld \n",count,*f_pos);
+
+    ret = mutex_lock_interruptible(&dev_struct->aesdchar_mutex);
+    if (ret < 0) {
+        printk(KERN_ERR"Mutex lock failed \n");
+        return ret;
+    }
+       
+    device_buffer = (char *)kmalloc(sizeof(char) * count, GFP_KERNEL);
```

Comment:
@sujoyray-ucolorado does this block and the one below it need to be inside the lock?  What is `&dev_struct->aesdchar_mutex` locking?

Reply:
Agreed -  further optimization can be done. This is not a bug though :-)